src/HistoryQuery/HistoryQueryEngine.class.js

```diff
@@ -0,0 +1,178 @@
+   * Creates CronJobs based on the ScanModes and starts them.
+   *
+   * @param {boolean} safeMode - Whether to start in safe mode
+   * @return {void}
+   */
+  async start(safeMode = false) {
+    this.logger.info('Starting HistoryQueryEngine')
+    if (safeMode || this.check) {
+      this.logger.warn('HistoryQuery is not executed in safe mode')
+      return
+    }
+    this.runNextHistoryQuery()
+  }
+
+  /**
+   * Gracefully stop every Timer, Protocol and Application
+   * @return {Promise<void>} - The stop promise
+   */
+  async stop() {
+    if (this.historyQuery) {
+      await this.historyQuery.stop()
+    }
+  }
+
+  /**
+   * Get the next history query config
+   * @returns {object|null} - The next history query config
+   */
+  getNextHistoryQueryConfig() {
+    const historyQueryConfigs = this.configService.getActiveHistoryQueryConfiguration()
+    const activeHistoryQueryConfigs = historyQueryConfigs.filter((historyQueryConfig) => historyQueryConfig.enabled && !historyQueryConfig.paused)
+
+    const ongoingHistoryQueries = activeHistoryQueryConfigs.filter(
+      (historyQueryConfig) => [HistoryQuery.STATUS_EXPORTING, HistoryQuery.STATUS_IMPORTING].includes(historyQueryConfig.status),
+    )
+    if (ongoingHistoryQueries.length > 0) {
+      return ongoingHistoryQueries.sort((a, b) => a.order - b.order)[0]
+    }
+
+    const pendingHistoryQueries = activeHistoryQueryConfigs.filter((historyQueryConfig) => historyQueryConfig.status === HistoryQuery.STATUS_PENDING)
+    if (pendingHistoryQueries.length > 0) {
+      return pendingHistoryQueries.sort((a, b) => a.order - b.order)[0]
+    }
+
+    return null
+  }
+
+  /**
+   * Run the next history query
+   * @returns {void}
+   */
+  runNextHistoryQuery() {
+    const historyQueryConfig = this.getNextHistoryQueryConfig()
+    if (historyQueryConfig) {
+      this.logger.info(`Preparing to start HistoryQuery: ${historyQueryConfig.id}`)
+
+      const { southConfig, northConfig } = this.configService.getConfig()
+      const dataSourceToUse = southConfig.dataSources.find((dataSource) => dataSource.id === historyQueryConfig.southId)
+      if (!dataSourceToUse) {
+        this.logger.error(`Invalid dataSource ${historyQueryConfig.southId} for HistoryQuery ${historyQueryConfig.id}`)
+      }
+      const applicationToUse = northConfig.applications.find((application) => application.id === historyQueryConfig.northId)
+      if (!applicationToUse) {
+        this.logger.error(`Invalid application ${historyQueryConfig.southId} for HistoryQuery ${historyQueryConfig.id}`)
+      }
+
+      if (dataSourceToUse && applicationToUse) {
+        this.historyQuery = new HistoryQuery(this, this.logger, historyQueryConfig, dataSourceToUse, applicationToUse)
+        this.historyQuery.start()
+      }
+    } else {
+      this.logger.warn('No HistoryQuery to execute')
+    }
+  }
+
+  /**
+   * Get cache folder
+   * @return {string} - The cache folder
+   */
+  getCacheFolder() {
```

Comment:
isn't it simpler to use the variable directly than to create this method?

Reply:
It is a method whit different implementations for OIBusEngine and HistoryQueryEngine.